Add script to generate metadata for use in compatibility testing.

alpearce opened the pull request:
(Also check in generated metadata.)

We want to be explicit about making changes to rust-tuf that change the metadata it generates. go-tuf accomplishes this by containing checked-in keys and metadata generated by go-tuf and, in a unit test, asserts that it produces byte-for-byte equivalent metadata. That way, changes to the library that change generated metadata must include corresponding changes to the checked-in metadata to indicate the change in behavior. 

This patch includes the metadata generator and the first set of generated metadata. It uses the same keys as the go-tuf generator so that this metadata can also be used for comparison across implementations. 

Feel free to go wild on the review; this is the biggest hunk of Rust I've written and I'd appreciate a lot of comments.

erickt commented on tests/metadata/generate.rs:
```diff
@@ -0,0 +1,262 @@
+use tuf::repository::{FileSystemRepository, FileSystemRepositoryBuilder, Repository};
+
+// TODO parametrize this better/find it within the build?
+const KEYS_PATH: &str = "./keys.json";
```
This should be okay for now. I've been thinking about re-laying out the directory structure, so we could handle improving discovery of the keys.json file then.

alpearce replied:
Ack.

erickt commented on tests/metadata/generate.rs:
```diff
@@ -0,0 +1,262 @@
+    let mut contents = String::new();
+    f.read_to_string(&mut contents)
+        .expect("failed to read keys");
+    let keys: TestKeys = serde_json::from_str(&contents).expect("serde failed");
```
You don't need to read this into a string. This should work:

```
let keys: TestKeys = serde_json::from_reader(f).expect("serde failed");
```

alpearce replied:
Done.

erickt commented on tests/metadata/generate.rs:
```diff
@@ -0,0 +1,262 @@
+        .join((step - 1).to_string())
+        .join("repository");
+    let dst = Path::new(dir).join(step.to_string());
+    Command::new("cp")
```
Could you use `/bin/cp`, just to make sure we don't pick up any old cp executable in the current directory? It should be okay if this only is runnable on a posix-y system, since this is a test script.

alpearce replied:
Done.

erickt commented on tests/metadata/generate.rs:
```diff
@@ -0,0 +1,262 @@
+    keys
+}
+
+fn copy_repo(dir: &str, step: u8) {
```
Generally speaking I'd prefer this call some rust function to copy the directories, since we might not guarantee that `cp` is available. Eventually I'd like to turn this into a test library, so that we can add a test that will complain if we accidentally change the metadata format without regenerating the test metadata, like we're doing in [go-tuf](https://fuchsia.googlesource.com/third_party/go-tuf/+/refs/heads/master/client/interop_test.go#32).

Unfortunately, I'm guessing you discovered there's no standard library way for copying directories. We might be able to pull in a crate for this ([fs_extra](https://crates.io/crates/fs_extra) seems popular, but we need to check out it's quality), but for now I say we punt for now and solve this later for when we create this identity test.

So in short, could you add a `// TODO replace this with pure rust library so it's portable across OSes`?

alpearce replied:
Added TODO.

erickt commented on tests/metadata/generate.rs:
```diff
@@ -0,0 +1,262 @@
+}
+
+// adds a target and updates the non-root metadata files.
+async fn add_target(
```
More evidence we need a [helper library for generating metadata](https://github.com/heartsucker/rust-tuf/issues/232).

cavokz replied:
@alpearce is it ok for you if I attempt at the library part or is it something you would like to do?

alpearce replied:
Fine with me!

erickt commented on tests/metadata/generate.rs:
```diff
@@ -0,0 +1,262 @@
+
+async fn generate_repos(dir: &str, consistent_snapshot: bool) -> tuf::Result<()> {
+    // Create initial repo.
+    // TODO: @erickt, this impl is very tied to the json keys file. Do we need it to be more flexible?
```
This should be fine for now.

alpearce replied:
Ack.

erickt commented on tests/metadata/generate.rs:
```diff
@@ -0,0 +1,262 @@
+            .unwrap();
+        copy_repo(dir, i);
+
+        let root_signer: Option<PrivateKey> = match r {
```
Nit: I think you can leave off `Option<PrivateKey>`.

alpearce replied:
Done.

erickt reviewed (changes requested):
A few things other things:

* First, it looks like you hit a [test failure](https://github.com/heartsucker/rust-tuf/pull/237/checks?check_run_id=315130477). You can ignore the appveyor, that's [going away](https://github.com/heartsucker/rust-tuf/issues/235).
* Darn, I thought I added a mode FileSystemBuilder that lets us pretty print the metadata. That makes it much easier to review. I filed https://github.com/heartsucker/rust-tuf/issues/238 to track implementing this.

alpearce commented:
Updated:
  - @erickt's comments
  - use a constant expiration so metadata will be static
  - regenerate metadata with the above

heartsucker commented:
Nice! We should fix #238 before merging this one otherwise reviews are going to be impossible.

alpearce commented:
Borrowed the "linkify-metadata" script from go-tuf and regenerated symlinked metadata.

erickt committed:
Add a json pretty printing interchange

This adds an interchange called `JsonPretty`, which makes it easier for
humans to review metadata changes. It is identical to
`tuf::interchange::Json`, except for the implementation of `to_writer`.
Instead of writing out the canonical form, it writes out a pretty printed
form of the data.

In addition, it changes `SnapshotMetadataBuilder::insert_metadata_with_path`
and `TimestampMetadataBuilder::from_snapshot` to use `D::to_writer`. This
makes sure that the hash of these metadata files are computed from the
format written to disk, instead of the canonical format.

erickt commented:
A couple more things:

* I uploaded a pretty printing interchange in https://github.com/heartsucker/rust-tuf/pull/241, could you rebase your patch on top of this one and use it to regenerate the metadata? While you do that, can you squash all the patches together, then factor out the metadata into it's own patch? That should make it easier to review the metadata generation script from the actual metadata.
* I used your PR to test out that `JsonPretty` worked (thanks!) I was able to successfully parse step `0`, but I ran into a bug in step `1`. It is missing the `0` target. Instead, it should look like [this](https://github.com/heartsucker/rust-tuf/blob/develop/tests/interop/fuchsia-go-tuf-5527fe/consistent-snapshot-false/1/repository/targets.json).
* You don't need to do this now, but it'd be nice if we could migrate this script into a separate independent cargo toml. I'd be fine if you want us to do this in a future PR though to get this landed.

Commit:
Add script to generate metadata for use in compatibility testing.

Commit:
upload generated metadata

alpearce commented:
Prettified, squashed, separated metadata, fixed bug, added TODO in Cargo.toml.
